src/backend/wayland/window.rs

```diff
@@ -245,12 +246,30 @@ impl WindowHandle {
         }
     }
 
-    pub fn request_timer(&self, _deadline: std::time::Instant) -> TimerToken {
-        todo!()
+    pub fn request_timer(&self, deadline: std::time::Instant) -> TimerToken {
+        let props = self.properties();
+        let props = props.borrow();
+        let window_id = WindowId::new(&props.wayland_window);
+        let token = TimerToken::next();
+        props
+            .loop_handle
+            .insert_source(
+                Timer::from_deadline(deadline),
+                move |_deadline, _, state| {
+                    let window = state.windows.get_mut(&window_id);
+                    if let Some(window) = window {
+                        window.handler.timer(token);
+                    }
+                    // In theory, we could get the `timer` request to give us a new deadline
+                    TimeoutAction::Drop
+                },
+            )
+            .expect("could add a timer loop");
```

Comment:
couldn't?

Reply:
Also, I realize the API wasn't designed for it, but I think we should avoid panicking as much as possible. I sort of expect failures like this to be not really recoverable, but at least the application should be able to save data before bailing out...

Reply:
We expected that we `could` - I believe that you `expect` what you expected, not what happened.

As far as I can tell you going through the control flow, this is definitely true anyway - i.e. could be an unwrap

Reply:
> I believe that you `expect` what you expected, not what happened

My first reaction was: "aaah, that finally makes sense", I was always confused by the name `expect`. But then the actual message printed to the terminal would be confusing, because it doesn't put it in the context (such as "it was expected that ..."). 

Turns out there are [guidelines on expect messages](https://doc.rust-lang.org/std/error/index.html#common-message-styles). And I remain confused why the method is called `expect`.

Reply:
But yeah, if panic shouldn't actually be possible, I guess `unwrap` communicates it better to other people reading the code

Reply:
Panicking as rarely as possible is something I strongly support in general. That said, I don't view it as a hard requirement for early work on features. So the panic could be removed during some later panic reduction work.

As for the panic message, none of our existing messages use this style. We mostly have descriptions of what failed, which I think is fine. We also have a few describing what should have happened. But yeah, none that describe what could happen instead.

I also don't really like the use of the keyword *should* as written in the guidelines, because it conflicts with the [classic RFC definition](https://datatracker.ietf.org/doc/html/rfc2119) of *must* vs *should* where *should* means actually optional. With `expect` it's clearly not optional.

As such I'd change it to either e.g. *`failed to add timer`* or *`timer adding must succeed`*.